src/modules/quick_api_module.f90

```diff
@@ -0,0 +1,688 @@
+    iatm = atomic_numbers(i)
+    
+    do j=1, natm_type
+      if(atm_type_id(j) .eq. iatm) new_atm_type = .false.
+    enddo
+
+    if(new_atm_type) then
+      atm_type_id(natm_type) = iatm
+      natm_type = natm_type+1
+    endif
+
+  enddo  
+
+  natm_type = natm_type-1   
+
+end subroutine get_atom_types
+
+! returns quick qm energy
+subroutine get_quick_energy(coords, xt_chg_crd, energy)
+
+  implicit none
+
+  double precision, intent(in)  :: coords(3,quick_api%natoms)
+  double precision, intent(in)  :: xt_chg_crd(4,quick_api%nxt_ptchg)
+  double precision, intent(out) :: energy
+
+  ! assign passed parameter values into quick_api struct
+  quick_api%coords         = coords
+  quick_api%xt_chg_crd     = xt_chg_crd
+
+  call run_quick(quick_api)
+
+  ! send back total energy and charges 
+  energy = quick_api%tot_ene
+
+end subroutine get_quick_energy
+
+
+! calculates and returns energy, gradients and point charge gradients
+subroutine get_quick_energy_gradients(coords, xt_chg_crd, &
+           energy, gradients, ptchg_grad)
+
+  implicit none
+
+  double precision, intent(in)    :: coords(3,quick_api%natoms)
+  double precision, intent(in)    :: xt_chg_crd(4,quick_api%nxt_ptchg)
+  double precision, intent(out)   :: energy
+  double precision, intent(out)   :: gradients(3,quick_api%natoms)
+  double precision, intent(inout) :: ptchg_grad(3,quick_api%nxt_ptchg)
+
+  ! assign passed parameter values into quick_api struct
+  quick_api%coords         = coords
+  if(quick_api%nxt_ptchg>0) then 
+    quick_api%xt_chg_crd = xt_chg_crd
+    quick_api%ptchg_grad = ptchg_grad
+  endif
+
+  call run_quick(quick_api)
+
+  ! send back total energy, gradients and point charge gradients
+  energy     = quick_api%tot_ene
+  gradients     = quick_api%gradient
+
+  if(quick_api%nxt_ptchg>0) ptchg_grad = quick_api%ptchg_grad
+
+end subroutine get_quick_energy_gradients
+
+
+! runs quick, partially resembles quick main program
+subroutine run_quick(self)
+
+  use quick_timer_module
+  use quick_method_module, only : quick_method
+  use quick_files_module
+  use quick_calculated_module, only : quick_qm_struct
+  use quick_gridpoints_module, only : quick_dft_grid, deform_dft_grid
+#ifdef MPIV
+  use quick_mpi_module
+#endif
+
+  implicit none
+
+  type(quick_api_type), intent(inout) :: self
+  integer :: ierr, i, j, k
+  logical :: failed = .false.
+
+  ! print step into quick output file
+  call print_step(self)  
+
+  ! if dft is requested, make sure to delete dft grid variables from previous 
+  ! the md step before proceeding
+  if(( self%step .gt. 1 ) .and. quick_method%DFT) call deform_dft_grid(quick_dft_grid)
+
+  ! set molecular information into quick_molspec
+  call set_quick_molspecs(quick_api)
+
+  ! start the timer for initial guess
+  call cpu_time(timer_begin%TIniGuess)
+
+  ! we will reuse density matrix for steps above 1. For the 1st step, we should
+  ! read basis file and run SAD guess. 
+  if(self%firstStep .and. self%reuse_dmx) then
```

Comment:
This seems wrong. Should this have an `or` statement? Since if we don't reuse the density matrix, then we will need to do a new guess.

Reply:
No, I just want to reuse the density matrix computed from the first step. For the first step, self%firstStep and self%reuse_dmx are both true. We will compute density matrix for this case, save it and set the value of self%firstStep to .false. In the next step, we will reuse the computed density matrix and code paths within the if statement will not be executed.

Reply:
I still don't understand. At the first step we need to do a guess because we have no density matrix. At subsequent steps we can choose to do a new guess or reuse the density matrix from the previous step. I have to admit that I don't know what happens if `quick_method%SAD` is `.false.`.  I also assume that we need to assign basis functions only during first step. Then we need
```
if (self%firstStep) then
  call getMol()
end if

if (self%firstStep .or. (.not. self%reuse_dmx)) then
  if (quick_method%SAD) call getMolSad()
  ! What do we do if we don't do a SAD guess? 
  ! Where do we delete the density matrix?
end if

self%firstStep = .false.
```